pipeline_dp/budget_accounting.py

```diff
@@ -151,17 +158,49 @@ def scope(self,
 
         Args:
             weight: budget weight of all operations made within this scope as
-             compared to.
-            aggregation_scope: if True, this is a high-level scope which
-             corresponds to one aggregation.
+             compared to other scopes with the same parent scope.
+            aggregation_scope: if True, this is an aggregation scope.
+            Aggregation scopes are high-level scopes which corresponds to the
+             whole aggregation computation graph (e.g. the call of
+             DPEngine.aggregate()). Aggregation scopes can not include each
+             other.
 
         Returns:
             the scope that should be used in a "with" block enclosing the
             operations consuming the budget.
         """
         aggregation_index = None
-        if aggregation_scope:
+        if aggregation_scope and (self._n_aggregations or
+                                  self._aggregation_weights):
+            if self._inside_aggregation_scope:
+                raise ValueError("Aggregation scopes can not be nested.")
             aggregation_index = self._get_next_aggregation_index()
+            if self._n_aggregations:
+                if aggregation_index >= self._n_aggregations:
+                    raise ValueError("Exceeded the number of allowed "
+                                     "aggregations. If you need more update "
+                                     "'n_aggregations' in the constructor of "
+                                     "BudgetAccountant")
+                if weight != 1:
+                    raise ValueError(f"When 'n_aggregations' is set in the "
+                                     f"constructor of BudgetAccountant, all "
+                                     f"aggregation weights have to be 1, but "
+                                     f"weight={weight}.")
+            elif self._aggregation_weights != None:
+                if aggregation_index >= len(self._aggregation_weights):
+                    raise ValueError("Exceeded the number of allowed "
+                                     "aggregations. If you need more update "
+                                     "'aggregation_weights' in the constructor "
+                                     "of BudgetAccountant")
+                expected_weight = self._aggregation_weights[aggregation_index]
+                if weight != expected_weight:
+                    raise ValueError(
+                        f"The provided weight for the aggregation "
+                        f"with index={aggregation_index} is {weight}, "
+                        f"but 'aggregation_weights' in the constructor "
+                        "of BudgetAccountant contains {expected_weight}")
+            else:
+                assert False, "It should not happen."
```

Comment:
Is it idiomatic to throw an error like this instead of `raise XError("It should not happen.")` in Python?